### PR TITLE
Make dropWhile in Stream and LazyList stack-safe

### PR DESCRIPTION
**Pull request by NthPortal**
Addresses at least part of @SethTisue's concern in scala/collection-strawman#197.

### CODE REVIEW

**Comment by NthPortal** [comment]
I'm happy to add to this PR with more overridden methods; I'm just not sure exactly which methods are needed atm

**Comment by NthPortal** [comment]
The partest right now doesn't check for stack safety well enough - need to ensure that `head` and `tail` get evaluated

File: test/files/run/lazylist-dropwhile-stack-safe.scala [hunk 1]
```diff
@@ -0,0 +1,5 @@
+object Test extends App {
+  // Should not overflow the stack
+  Iterator.iterate(LazyList.from(0))(_.dropWhile(_ => false)).drop(10000).next
+  Iterator.iterate(  Stream.from(0))(_.dropWhile(_ => false)).drop(10000).next
```
Comment (NthPortal) [inline test/files/run/lazylist-dropwhile-stack-safe.scala]: doesn't check for stack safety well enough - need to ensure that head and tail get evaluated (for both `LazyList` and `Stream`)

Comment (NthPortal) [reply test/files/run/lazylist-dropwhile-stack-safe.scala]: Although the actual problem Seth mentioned doesn't seem to manifest with `LazyList` anyway, it's good to check that it works

**Comment by javax-swing** [comment]
@NthPortal I wonder if it's a valid test to check if a they use memory sharing.

```
val stream = Stream.of(1,2,3,4,5)
stream.drop(3) should be theSameInstanceAs stream.tail.tail.tail
```

```
val stream = Stream.of(1,2,3,4,5)
stream.dropWhile(_ <= 3) should be theSameInstanceAs stream.tail.tail.tail
```

```
val stream = LazyList.of(1,2,3,4,5)
stream.drop(3) should be theSameInstanceAs stream.tail.tail.tail
```
```
val stream = LazyList.of(1,2,3,4,5)
stream.dropWhile(_ <= 3) should be theSameInstanceAs stream.tail.tail.tail
```

**Comment by NthPortal** [comment]
@javax-swing while they *should* do that, that feels to me like specifying an internal behaviour for no reason; maybe that's just me though

**Comment by NthPortal** [comment]
ping @SethTisue

**Review by lrytz — APPROVED** [review]

**Comment by lrytz** [comment]
Thanks, @NthPortal!

**Pull request merged**